openvpn/mbedtls/crypto/cipheraead.hpp

```diff
@@ -0,0 +1,285 @@
+//    OpenVPN -- An application to securely tunnel IP networks
+//               over a single port, with support for SSL/TLS-based
+//               session authentication and key exchange,
+//               packet encryption, packet authentication, and
+//               packet compression.
+//
+//    Copyright (C) 2012-2017 OpenVPN Inc.
+//
+//    This program is free software: you can redistribute it and/or modify
+//    it under the terms of the GNU Affero General Public License Version 3
+//    as published by the Free Software Foundation.
+//
+//    This program is distributed in the hope that it will be useful,
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program in the COPYING file.
+//    If not, see <http://www.gnu.org/licenses/>.
+
+// Wrap the mbed TLS AEAD API.
+
+#ifndef OPENVPN_MBEDTLS_CRYPTO_CIPHERAEAD_H
+#define OPENVPN_MBEDTLS_CRYPTO_CIPHERAEAD_H
+
+#include <string>
+
+#include <mbedtls/gcm.h>
+#include <mbedtls/chachapoly.h>
+
+#include <openvpn/common/size.hpp>
+#include <openvpn/common/exception.hpp>
+#include <openvpn/common/likely.hpp>
+#include <openvpn/crypto/static_key.hpp>
+#include <openvpn/crypto/cryptoalgs.hpp>
+
+namespace openvpn
+{
+    namespace MbedTLSCrypto
+    {
+        class CipherContextAEAD
+        {
+            CipherContextAEAD(const CipherContextAEAD&) = delete;
+            CipherContextAEAD& operator=(const CipherContextAEAD&) = delete;
+
+            private:
+
+            bool initialized;
+            mbedtls_gcm_context gcm_ctx;
+            mbedtls_chachapoly_context chachapoly_ctx;
+            CryptoAlgs::Type crypto_alg;
+            
+            void (openvpn::MbedTLSCrypto::CipherContextAEAD::*aead_encrypt)(const unsigned char *input, unsigned char *output, size_t length, const unsigned char *iv, unsigned char *tag, const unsigned char *ad, size_t ad_len);
+            bool (openvpn::MbedTLSCrypto::CipherContextAEAD::*aead_decrypt)(const unsigned char *input, unsigned char *output, size_t length, const unsigned char *iv, const unsigned char *tag, const unsigned char *ad, size_t ad_len);
+
+            public:
+
+            OPENVPN_EXCEPTION(mbedtls_aead_error);
+
+            // mode parameter for constructor
+            
+            enum
+            {
+                MODE_UNDEF = MBEDTLS_OPERATION_NONE,
+                ENCRYPT = MBEDTLS_ENCRYPT,
+                DECRYPT = MBEDTLS_DECRYPT
+            };
+
+            // mbed TLS cipher constants
+      
+            enum
+            {
+                IV_LEN = 12,
+                AUTH_TAG_LEN = 16,
+                SUPPORTS_IN_PLACE_ENCRYPT = 1,
+            };
+
+#if 0
```

Comment:
Don't introduce something like an #if 0 without a comment

Reply:
This is class has been based on ciphergcm.hpp. And this statement is also found on that file.